### PR TITLE
Benchmarking and Evaluation Pipeline

### PR DESCRIPTION
**Pull request by arjun-banerjee**
**This is a new PR! The older one had files that I should've gitignored**

*High Level Description of changes:*
Wrote benchmarking tests for all the kernels we'd like to test (full list: [here](https://docs.google.com/document/d/1uM5L_1_sjcZi4r8XYpHAO4v2M0rbJzhv3bOp4ioOzBk/edit?tab=t.0#heading=h.jozb8x26681p)) and then created a pipeline to go from generation to benchmarking.

*Code Explanations"
All edits are in the evaluation/benchmarking/ directory.

The tests are in individual subfolders arranged by category. The test files end in `{category}_tests.py`. In general, the tests for each kernel function by making random vectors/matrixes and then comparing the outputs of the Pytorch and NKI functions. If they are the same, a 1 is returned; else, a 0 is returned.  The main functions of each of these files stores the results of the tests passing in a dictionary, which can be easily turned into a dashboard for visualization in the future.

A series of helper tests are created in `benchmarking_tools.py` -- the general process if that text is extracted from the LLM response (the `extract_kernel_from_llm_response` function is now modified to include more general extraction compared to the previous method), then replace the function name with the intended function name if it is not the same, the write the text into a python file.

The benchmarks are then run by running the `run_benchmarking.py` which turns all the LLM responses into python files using the benchmarking tools files, and then runs the tests defined in the `{category}_tests.py` files. It then saves the results in a JSON file.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: evaluation/benchmarking/benchmarking_tools.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    Args:
+        kernel_list (list): List of kernel names to process
+        kernel_name_type (str): Type of kernel being tested
+    """
+    # Create or clear the output file
+    output_file = f'{kernel_name_type}.py'
+    make_txt_file(output_file, '')
+    
+    # Add imports at the top of the file
+    append_txt_file(output_file, 'import torch\nimport torch.nn.functional as F\n\n')
```
Comment (EmilyWebber) [inline evaluation/benchmarking/benchmarking_tools.py:25]: Nice - this will also help if you ever want to explicitly integrate the kernels into another library, like NxD. Should also help when you want to generate different versions of those packages, and for different versions of NKI (updates, etc).

File: evaluation/benchmarking/edgeops/edgeops_tests.py [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+    match = torch.allclose(out_ref, out_nki, atol=1e-3, rtol=1e-2)
+    print("NKI and reference match!" if match else "NKI and reference differ")
+    return 1 if match else 0
+
+def test_torch_view_as_real(device):
+    """Test view_as_real (converts a complex tensor into a real tensor with extra dimension) between NKI and reference implementation."""
+    real = torch.randn((64, 128), device=device, dtype=torch.float32)
+    imag = torch.randn((64, 128), device=device, dtype=torch.float32)
+    x = torch.complex(real, imag)
+    out_nki = nki_view_as_real(x)
```
Comment (EmilyWebber) [inline evaluation/benchmarking/edgeops/edgeops_tests.py:89]: Cool! Did this one pass?

File: evaluation/benchmarking/elementwise/elementwise_tests.py [hunk 3]
```diff
@@ -0,0 +1,573 @@
...
+    """
+    Test elementwise inverse sine between NKI and PyTorch implementations.
+    
+    Args:
+        device: The device to run the test on (CPU/GPU/NPU)
+    
+    Returns:
+        int: Returns 1 if NKI and PyTorch results match, 0 otherwise
+    """
+    x_small = torch.rand((64, 128), dtype=torch.bfloat16, device=device) * 2 - 1  # Values between -1 and 1
```
Comment (EmilyWebber) [inline evaluation/benchmarking/elementwise/elementwise_tests.py:323]: Is there no first party way to set the range on torch tensors? Might give you more control and more ability to reset parameters on your tests. torch.clamp?

File: evaluation/benchmarking/products/products_tests.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+    
+    Args:
+        device: The device to run the test on (CPU/GPU/NPU)
+        
+    Returns:
+        int: Returns 1 if NKI and PyTorch results match, 0 otherwise
+    """
+    a = torch.rand((64, 128), dtype=torch.bfloat16, device=device)
+    b = torch.rand((128, 32), dtype=torch.bfloat16, device=device)
+    equation = "ij,jk->ik"
```
Comment (EmilyWebber) [inline evaluation/benchmarking/products/products_tests.py:210]: Ah there are many more Einsums to test out. This one is a good start.

File: evaluation/benchmarking/multielement/multielement_tests.py [hunk 5]
```diff
@@ -0,0 +1,740 @@
...
+def test_torch_round(device):
+    """Test rounding operation between NKI and PyTorch implementations.
+    
+    Args:
+        device: The device to run the test on (CPU/GPU/NPU)
+        
+    Returns:
+        int: Returns 1 if NKI and PyTorch results match, 0 otherwise
+    """
+    x_small = torch.rand((64, 128), dtype=torch.bfloat16, device=device) * 10 - 5
```
Comment (EmilyWebber) [inline evaluation/benchmarking/multielement/multielement_tests.py:238]: Why the sidecar ops on this one?

File: evaluation/benchmarking/multielement/multielement_tests.py [hunk 6]
```diff
@@ -0,0 +1,740 @@
...
+    Args:
+        device: The device to run the test on (CPU/GPU/NPU)
+        
+    Returns:
+        int: Returns 1 if NKI and PyTorch results match, 0 otherwise
+    """
+    x_small = torch.rand((64, 128), dtype=torch.bfloat16, device=device)
+    condition = x_small > 0.5
+    other = torch.zeros_like(x_small)
+    output_small = nki_where(condition, x_small, other)
```
Comment (EmilyWebber) [inline evaluation/benchmarking/multielement/multielement_tests.py:326]: Good to see some selection logic

**Review by EmilyWebber — APPROVED** [review]
This is looking great. Thank you for the cleanliness, simplicity, and consistency of the test suite!

You might want to consider adding some parameters to the workhorse test functions themselves. Things like input and output shape sizes especially can get challenging, and you might want to know if the kernels pass at p-dim of 64 but fail at 256, etc. It would also be useful to see data types as a parameter (do we pass at bf16 but not FP32, etc).
 
Do you have any functions in mind to check key LLM ops? Distance between sequences is usually calculated with KL divergence. Perplexity is key for pretraining, etc. 

Lastly you might consider adding some tests around data movement and manipulations a touch more generally. Can you accelerate extractions from matrices, transformations around them, interpolations between them, etc?
